library/std/src/sys/pal/windows/stdio.rs

```diff
@@ -19,13 +17,9 @@ pub struct Stdin {
     incomplete_utf8: IncompleteUtf8,
 }
 
-pub struct Stdout {
-    incomplete_utf8: IncompleteUtf8,
-}
+pub struct Stdout {}
```

Comment:
The reason for the removal of incomplete UTF-8 handling at the end of the string is not clear to me from the commit description. Why was that removed?

Reply:
Because it now simply truncates the write to remove incomplete UTF-8 from the end and instead leaves the buffering to buffer types, i.e. `LineWriter` in this case.

Reply:
A `LineWriter` will flush an incomplete line if its buffer capacity is exceeded. If that happens, the output must support partial UTF-8 writes, or non-ASCII characters might get lost or replaced with the replacement character.

Reply:
That can only result in broken UTF-8 if the user writes incomplete UTF-8 to `LineWriter` themselves.

Reply:
I see, digging through the source code, `BufWriter` makes sure to not split writes that the user issued.

What is the motivation for truncating invalid UTF-8 at the end of the string?

All else being equal, I'd rather expect the previous behavior, that I can construct UTF-8 output byte-by-byte.

Reply:
Rather than having a secret stack buffer that can't be inspected or flushed, I'd strongly prefer buffering be done at a higher level. It's also a lot of added complexity for an edge case where the better solution is to set the console code page.

Reply:
In any case, if that behavior is wanted, it should probably be documented in the commit message so that it is clear to future readers that this change was on purpose.

Rather than ignoring trailing invalid UTF-8, I think it'd be better to replace it with a replacement character so that it becomes clear that something was removed.

Reply:
> Rather than ignoring trailing invalid UTF-8, I think it'd be better to replace it with a replacement character so that it becomes clear that something was removed.

That's what happens in this code. No bytes are ever lost. Either the caller is informed that less bytes were written than were provided or, if there is only an incomplete code point, then the that is written to the console (which will be converted to replacement characters when lossy translating to UTF-16).

Reply:
Ah yes, that makes sense. I didn't realize the caller would be informed by the return value of `write`.

What is the motivation for special casing trailing invalid UTF-8? It seems to increase the code complexity a little as well, and is not necessary for std's own use cases.

Is it for supporting a potential non-std buffered writer?

Reply:
Sure, it could be removed. `Stderr` is not buffered by us though and there have been proposals for unbuffered stdout.